### PR TITLE
IGVF-726-virtual-to-sample-and-donor

### PR DESCRIPTION
**Pull request by KAndreeva**
Added virtual property (boolean) to sample and donor.

### CODE REVIEW

File: src/igvfd/schemas/sample.json [hunk 1]
```diff
@@ -115,6 +115,12 @@
             "description": "Detail for sample sorted into fractions capturing information about sorting.",
             "comment": "For example, FACS 10% expression of CTCF-GFP",
             "type": "string"
+        },
+        "virtual": {
+            "title": "Virtual",
+            "description": "Sample is virtusal.",
```
Comment (gabdank) [inline src/igvfd/schemas/sample.json:121]: typo "virtusal" -> virtual

File: src/igvfd/schemas/donor.json [hunk 2]
```diff
@@ -81,6 +81,12 @@
                 "type": "string",
                 "linkTo": "PhenotypicFeature"
             }
+        },
+        "virtual": {
+            "title": "Virtual",
+            "description": "Donor is virtusal.",
```
Comment (gabdank) [inline src/igvfd/schemas/donor.json:87]: typo "virtusal" -> virtual

**Review by gabdank — CHANGES_REQUESTED** [review]
I think virtual should be a REQUIRED field

File: src/igvfd/schemas/donor.json (hunk 2 above)
Comment (KAndreeva) [reply src/igvfd/schemas/donor.json]: fixed

File: src/igvfd/schemas/sample.json (hunk 1 above)
Comment (KAndreeva) [reply src/igvfd/schemas/sample.json]: fixed. Thank you!

File: src/igvfd/tests/test_schema_human_donor.py [hunk 3]
```diff
@@ -77,3 +77,23 @@ def test_patch_related_donors(human_donor, parent_human_donor_1, testapp):
...
+        human_donor['@id'],
+        {'virtual': False})
+    assert res.status_code == 200
+    res = testapp.patch_json(
+        human_donor['@id'],
+        {'virtual': ''}, expect_errors=True)
+    assert res.status_code == 422
+
+
+def test_read_virtual(human_male_donor, testapp):
```
Comment (gabdank) [inline src/igvfd/tests/test_schema_human_donor.py:97]: I am not sure what is this test?

File: src/igvfd/tests/test_schema_primary_cell.py [hunk 4]
```diff
@@ -91,3 +91,23 @@ def test_sorted_fraction_detail_dependency(testapp, primary_cell):
...
+        primary_cell['@id'],
+        {'virtual': False})
+    assert res.status_code == 200
+    res = testapp.patch_json(
+        primary_cell['@id'],
+        {'virtual': ''}, expect_errors=True)
+    assert res.status_code == 422
+
+
+def test_read_virtual(primary_cell_v7, testapp):
```
Comment (gabdank) [inline src/igvfd/tests/test_schema_primary_cell.py:111]: i don't think we need this test

File: src/igvfd/tests/test_upgrade_in_vitro_system.py [hunk 5]
```diff
@@ -58,3 +58,9 @@ def test_in_vitro_system_upgrade_8_9(upgrader, in_vitro_system_v8):
 def test_in_vitro_system_upgrade_9_10(upgrader, in_vitro_system_v9):
     value = upgrader.upgrade('in_vitro_system', in_vitro_system_v9, current_version='9', target_version='10')
     assert value['schema_version'] == '10'
+
+
+def test_in_vitro_system_upgrade_10_11(upgrader, in_vitro_system_v10):
+    value = upgrader.upgrade('in_vitro_system', in_vitro_system_v10, current_version='10', target_version='11')
+    assert value['schema_version'] == '11'
+    assert 'virtual' in value
```
Comment (gabdank) [inline src/igvfd/tests/test_upgrade_in_vitro_system.py:66]: assert value in virtual

File: src/igvfd/tests/test_upgrade_rodent_donor.py [hunk 6]
```diff
@@ -37,3 +37,9 @@ def test_rodent_donor_upgrade_6_7(upgrader, rodent_donor_v6_with_parents, parent
                              current_version='6', target_version='7')
     assert 'parents' not in value
     assert value['notes'] == 'This is a note.  parents: ' + parent_rodent_donor_1['@id']
+
+
+def test_rodent_donor_upgrade_7_8(upgrader, rodent_donor_v7):
+    value = upgrader.upgrade('rodent_donor', rodent_donor_v7, current_version='7', target_version='8')
+    assert value['schema_version'] == '8'
+    assert 'virtual' in value
```
Comment (gabdank) [inline src/igvfd/tests/test_upgrade_rodent_donor.py:45]: assert value of virtual

File: src/igvfd/tests/test_upgrade_technical_sample.py [hunk 7]
```diff
@@ -31,3 +31,9 @@ def test_technical_sample_upgrade_5_6(upgrader, technical_sample_v5):
     value = upgrader.upgrade('technical_sample', technical_sample_v5, current_version='5', target_version='6')
     assert value['schema_version'] == '6'
     assert value['sorted_fraction_detail'] == 'Default upgrade text: please add more details about sorted_fraction, see sample.json for description.'
+
+
+def test_technical_sample_upgrade_6_7(upgrader, technical_sample_v6):
+    value = upgrader.upgrade('technical_sample', technical_sample_v6, current_version='6', target_version='7')
+    assert value['schema_version'] == '7'
+    assert 'virtual' in value
```
Comment (gabdank) [inline src/igvfd/tests/test_upgrade_technical_sample.py:39]: assert value of virtual

File: src/igvfd/tests/test_upgrade_tissue.py [hunk 8]
```diff
@@ -85,3 +85,9 @@ def test_tissue_upgrade_11_12(upgrader, tissue_v11):
     assert value['schema_version'] == '12'
     assert 'taxa' not in value
     assert value['notes'] == 'Previous taxa: Homo sapiens will now be calculated.'
+
+
+def test_tissue_upgrade_12_13(upgrader, tissue_v12):
+    value = upgrader.upgrade('tissue', tissue_v12, current_version='12', target_version='13')
+    assert value['schema_version'] == '13'
+    assert 'virtual' in value
```
Comment (gabdank) [inline src/igvfd/tests/test_upgrade_tissue.py:93]: assert value of virtual

File: src/igvfd/tests/test_upgrade_whole_organism.py [hunk 9]
```diff
@@ -98,3 +98,9 @@ def test_whole_organism_upgrade_12_13(upgrader, whole_organism_v12):
     assert value['schema_version'] == '13'
     assert 'taxa' not in value
     assert value['notes'] == 'Previous taxa: Homo sapiens will now be calculated.'
+
+
+def test_whole_organism_upgrade_13_14(upgrader, whole_organism_v13):
+    value = upgrader.upgrade('whole_organism', whole_organism_v13, current_version='13', target_version='14')
+    assert value['schema_version'] == '14'
+    assert 'virtual' in value
```
Comment (gabdank) [inline src/igvfd/tests/test_upgrade_whole_organism.py:106]: assert value of virtual

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/tests/test_schema_human_donor.py (hunk 3 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_schema_human_donor.py]: This test was to confirm that the default value for the schema property virtual was working correctly.  The test has been removed.

File: src/igvfd/tests/test_schema_primary_cell.py (hunk 4 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_schema_primary_cell.py]: This test was to confirm that the default value for the schema property virtual was working correctly.  The test has been removed.

File: src/igvfd/tests/test_upgrade_human_donor.py [hunk 10]
```diff
@@ -65,3 +65,9 @@ def test_human_donor_upgrade_7_8(upgrader, human_donor_v7_with_parents, parent_h
         'donor': parent_human_donor_2['@id'],
         'relationship_type': 'parent'
     } in value['related_donors']
+
+
+def test_human_donor_upgrade_8_9(upgrader, human_donor_v8):
+    value = upgrader.upgrade('human_donor', human_donor_v8, current_version='8', target_version='9')
+    assert value['schema_version'] == '9'
+    assert 'virtual' in value
```
[the remark this answers is not in the document]
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_human_donor.py]: Assert comparison is updated to compare the property to default property value of False.

File: src/igvfd/tests/test_upgrade_in_vitro_system.py (hunk 5 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_in_vitro_system.py]: Assert comparison is updated to compare the property to default property value of False.

File: src/igvfd/tests/test_upgrade_rodent_donor.py (hunk 6 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_rodent_donor.py]: Assert comparison is updated to compare the property to default property value of False.

File: src/igvfd/tests/test_upgrade_technical_sample.py (hunk 7 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_technical_sample.py]: Assert comparison is updated to compare the property to default property value of False.

File: src/igvfd/tests/test_upgrade_tissue.py (hunk 8 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_tissue.py]: Assert comparison is updated to compare the property to default property value of False.

File: src/igvfd/tests/test_upgrade_whole_organism.py (hunk 9 above)
Comment (KAndreeva) [reply src/igvfd/tests/test_upgrade_whole_organism.py]: Assert comparison is updated to compare the property to default property value of False.

**Review by gabdank — DISMISSED** [review]

**Review by gabdank — APPROVED** [review]

**Review by gabdank — APPROVED** [review]

**Pull request merged**